code/modules/mob/living/carbon/carbon_update_icons.dm

```diff
@@ -224,76 +224,70 @@
 		var/obj/item/bodypart/BP = X
 		BP.original_owner = WEAKREF(src)
 
-/mob/living/carbon/proc/update_body_parts()
-	//CHECK FOR UPDATE
-	var/oldkey = icon_render_key
-	icon_render_key = generate_icon_render_key()
-	if(oldkey == icon_render_key)
+/mob/living/carbon/proc/update_body_parts(var/update_limb_data)
+	//Check the cache to see if it needs a new sprite
+	update_damage_overlays()
+	var/list/needs_update = list()
+	var/limb_count_update = FALSE
+	for(var/obj/item/bodypart/BP as anything in bodyparts)
+		BP.update_limb(is_creating = update_limb_data) //Update limb actually doesn't do much, get_limb_icon is the cpu eater.
+		var/old_key = icon_render_keys?[BP.body_zone]
+		icon_render_keys[BP.body_zone] = (BP.is_husked) ? generate_husk_key(BP) : generate_icon_key(BP)
+		if(!(icon_render_keys[BP.body_zone] == old_key))
+			needs_update += BP
+
+
+	var/list/missing_bodyparts = get_missing_limbs()
+	if(((dna ? dna.species.max_bodypart_count : 6) - icon_render_keys.len) != missing_bodyparts.len)
+		limb_count_update = TRUE
+		for(var/X in missing_bodyparts)
+			icon_render_keys -= X
+
+	if(!needs_update.len && !limb_count_update)
 		return
 
 	remove_overlay(BODYPARTS_LAYER)
 
-	for(var/X in bodyparts)
-		var/obj/item/bodypart/BP = X
-		BP.update_limb()
-
-	//LOAD ICONS
-	if(limb_icon_cache[icon_render_key])
-		load_limb_from_cache()
-		return
-
 	//GENERATE NEW LIMBS
 	var/list/new_limbs = list()
-	var/draw_features = !HAS_TRAIT(src, TRAIT_INVISIBLE_MAN)
-	for(var/X in bodyparts)
-		var/obj/item/bodypart/BP = X
-		new_limbs += BP.get_limb_icon(draw_external_organs = draw_features)
+	for(var/obj/item/bodypart/BP as anything in bodyparts)
+		if(BP in needs_update)
+			var/bp_icon = BP.get_limb_icon()
+			new_limbs += bp_icon
+			limb_icon_cache[icon_render_keys[BP.body_zone]] = bp_icon
+		else
+			new_limbs += limb_icon_cache[icon_render_keys[BP.body_zone]]
+
 	if(new_limbs.len)
 		overlays_standing[BODYPARTS_LAYER] = new_limbs
-		limb_icon_cache[icon_render_key] = new_limbs
 
 	apply_overlay(BODYPARTS_LAYER)
-	update_damage_overlays()
 
 
 
-/////////////////////
-// Limb Icon Cache //
-/////////////////////
+/////////////////////////
+// Limb Icon Cache 2.0 //
+/////////////////////////
+//Updated by Kapu#1178
```

Comment:
Git blame will say who last updated it, no need to include comments

Reply:
Also I have been told by Mothblocks that doing these big comment blocks like this is to be avoided, but having one at the top of the file is usually an exception if one exists